allow token iteration in range for

shaneasd opened the pull request:
Branched from https://github.com/danmar/cppcheck/pull/3014/files#r559145415

Adds a mechanism to allow iteration over tokens of the form
```for( Token* x : tok->toend())```

shaneasd commented on lib/token.h:
```diff
@@ -187,6 +187,30 @@ class CPPCHECKLIB Token {
     explicit Token(TokensFrontBack *tokensFrontBack = nullptr);
     ~Token();
 
+    struct TokenEnumeration
```
@danmar regarding the question of whether this should be in Token, mTokensFrontBack is a private member of Token so if this class isn't part of Token I would have to expose mTokensFrontBack somehow.

pfultz2 replied:
This should probably be templated so it can iterate over both `Token* and `const Token*`(probably as `template<class T, REQUIRES("T must be a Token class", std::is_convertible<T*, const Token*>)> struct TokenEnumeration` so the template errors are clearer).

Also, this is considered a range in C++. It seems like `TokenRange` would be a better name, but its up to you.

danmar replied:
I like the name `TokenRange`.

I am not sure if we need `Token*`.. I spontanously feel it would not be extremely useful. If we iterate the non-const tokens and modify the list somehow then a range loop might not be a good idea. Probably it's not very useful in Tokenizer. In createAst I don't think we could use this neither. Would that be useful in ValueFlow?

shaneasd replied:
Even if we don't want a non-const version (which I also couldn't think of a use for) it might be worth explicitly disabling for improved compiler diagnostics as suggested. At that point it might be simpler to implement rather than not.

shaneasd replied:
unfortunately the example given for the REQUIRES macro prevents forward declaration of the type which I think is a problem

pfultz2 replied:
> Would that be useful in ValueFlow?

> If we iterate the non-const tokens and modify the list somehow then a range loop might not be a good idea. 

We should probably extend the `invalidContainerLoop` to detect this issue with tokens.

Yes, its quite common in ValueFlow to use `Token*` because we need to modify the values in the token. However, we dont modify the tokens order in ValueFlow.

Of course, it would be nice if we could get rid of this, as it could simplify a lot of things such as extra overloads for both const and non-const in astutils and reusing the analyzers outside of valueflow. One way to do this is to make the values a pointer and then in valueflow use a map from the token to values pointer(ie `std::map<const Token*, std::shared_ptr<std::list<ValueFlow::Value>>>`) this way valueflow can update the values using `const Token*`. 

Although, for the short-term I dont see that happening, so it seems it would be good to have both.

pfultz2 replied:
> unfortunately the example given for the REQUIRES macro prevents forward declaration of the type which I think is a problem

We should template a base class and then inherit it for const and non-const versions:

```cpp
template<class T, REQUIRES("T must be a Token class", std::is_convertible<T*, const Token*>)>
struct TokenRangeBase {
    // ...
};

struct ConstTokenRange : TokenRangeBase<const Token> {
    using TokenRangeBase<const Token>::TokenRangeBase<const Token>;
};

struct TokenRange : TokenRangeBase<Token> {
    using TokenRangeBase<Token>::TokenRangeBase<Token>;
};
```

That should simplify forward declarations.

shaneasd replied:
This is no longer true since I'm using nullptr rather than mTokensFrontBack

shaneasd replied:
good idea

pfultz2 commented:
I think this is a great idea. Maybe we could have a function `tok->until(end)` so we could iterate until a certain token.

pfultz2 commented on lib/token.h:
```diff
@@ -187,6 +187,30 @@ class CPPCHECKLIB Token {
 
+    struct TokenEnumeration
+    {
+        Token const* mt;
```
It should probably store the start and the end, so it can iterate over partial ranges.

shaneasd replied:
you're right. I have a local change that does that which I'll push some time soon.

pfultz2 commented on lib/token.h:
```diff
@@ -187,6 +187,30 @@ class CPPCHECKLIB Token {
+        struct TokenIterator
+        {
+            const Token* mt;
+            TokenIterator(const Token* t) : mt(t) {}
```
This needs a default constructor. Iterators are required to be default constructible(see [here](https://en.cppreference.com/w/cpp/named_req/ForwardIterator)).

pfultz2 commented on lib/token.h:
```diff
@@ -187,6 +187,30 @@ class CPPCHECKLIB Token {
+        Token const* mt;
+        TokenEnumeration(Token const* t) : mt(t) {}
+
+        struct TokenIterator
```
For compatibility with C++11 algorithms, you should add these typedefs:

```cpp
using iterator_category = std::forward_iterator_tag;
using value_type = Token*;
using difference_type = std::ptrdiff_t;
using pointer = Token*;
using reference = Token*;
```

shaneasd replied:
it looks wrong to me that value_type, pointer and reference would all be the same type. Shouldn't pointer be either Token** or void?

pfultz2 replied:
The `reference` type should be the result of `operator*` and `pointer` should be the result of `operator->`. You are missing `operator->` in your iterator, but depending on how you define `->` this typedef should be set accordingly.

shaneasd replied:
How would you define ```->```? Normally I would have it return a pointer to the thing being iterated over which in this case would be ```T**```. That  doesn't seem very useful though. You could make it ```T*``` instead but that would mean effectively a double dereference which seems bad as well. I think the best option might be to make pointer ```void``` to reflect the fact that there is no operator->

pfultz2 replied:
It would be useful to have a `->` especially when used with algorithms. Something like:

```cpp
...
auto it = std::find_if(start, last, predicate);
if (it != last) {
    return it->values();
}
...
```

For such an example to work we would need to return `T*`, I believe. If we return `T**` it would require an extra dereference to work(`operator->` will only drill down until it hits a pointer type, which means an extra dereference is necessary). Either way, I dont think its possible to return `T**` as it would create a dangling pointer. To avoid this, you could return proxy instead:

```cpp
        struct proxy {
            T* x;
            T* operator->() const { return x; }
        };
        proxy operator->() const { return proxy{mt}; }
```

But it seems like just returning `T*` is simpler.

pfultz2 replied:
Of course, you could leave it as `void` for now, and we can add `operator->` at a later time.

shaneasd replied:
I agree T* makes more sense than T** or proxy. My concern is that it->values makes it look like it as a Token* which it isn't.

danmar commented on lib/token.h:
```diff
@@ -187,6 +187,30 @@ class CPPCHECKLIB Token {
+        };
+
+        TokenIterator begin() { return TokenIterator(mt); }
+        TokenIterator end() { return TokenIterator(mt->mTokensFrontBack->back); }
```
Shouldn't the end token be nullptr if we want to loop through all tokens?

As @pfultz2 says, it would be very useful with partial ranges.

We have such code in many places:

    for (const Token *tok = someScope->bodyStart; tok != someScope->bodyEnd; tok = tok->next())

I would like to replace that with something like:

    for (const Token *tok: someScope->bodyTokens())

shaneasd replied:
I think you're probably right and that this contains a bug. I'll be adding some unit tests which hopefully will identify this bug.

shaneasd replied:
I've added tests and this was indeed a bug

shaneasd replied:
example of a partial range can be seen in the test ```partialEnumeration```. It looks like this
```
        const Token* expected = var.tokens()->tokAt(4);
        const Token* end = var.tokens()->tokAt(10);
        for (auto t : TokenRange<const Token>{ expected, end }) {
```

danmar reviewed (commented):
Good idea!

shaneasd commented:
PR is cleaned up and ready for review now. I've made changes based on the review of the draft. I moved functions/classes around into a more sensible arrangement so that dependencies can be a bit cleaner. I'm not totally sold on the templating. It concerns me that unless you're using toEnd you have to type ```const Token``` which seems a bit redundant (I think you might not have to with c++17 onwards but I haven't tested since I believe cppcheck must compile with c++11?). I've added unit tests which both test the expected behavior and demonstrate possible usages. To do this I had to make the assertion macros more powerful. Not sure why they were so limited before so this might be a mistake but I'm not sure how else to achieve the test behavior I'm after.

pfultz2 commented on lib/tokenrange.h:
```diff
@@ -0,0 +1,58 @@
+        using value_type = Token*;
+        using difference_type = std::ptrdiff_t;
+        using pointer = Token**;
+        using reference = Token*&;
```
This should be `T*` as this is matches the return of `operator*`.

pfultz2 commented on lib/tokenrange.h:
```diff
@@ -0,0 +1,58 @@
+    struct TokenIterator
+    {
+        using iterator_category = std::forward_iterator_tag;
+        using value_type = Token*;
```
This should also be `T*`.

pfultz2 commented on lib/token.cpp:
```diff
@@ -53,6 +54,15 @@ Token::~Token()
+TokenRange<const Token> Token::toEnd() const
+{
+    return TokenRange<const Token>(this, nullptr);
+}
```
What about another method `until`?

```cpp
TokenRange<const Token> Token::until(const Token* end) const
{
    return TokenRange<const Token>(this, end);
}
```

shaneasd replied:
I'm concerned that it would be ambiguous as to whether that would be inclusive or exclusive.

I think ```until``` would render toEnd largely redundant as I'd just as soon type ```until(nullptr)``` rather than ```toEnd```.

pfultz2 replied:
> I'm concerned that it would be ambiguous as to whether that would be inclusive or exclusive.

I would assume it would always be one past the end as its consistent with the standard library, but this might be better as a free function anyways.

pfultz2 replied:
We could add a `makeTokenRange` function(which should automatically deduce the type):

```cpp
template<class T, REQUIRES("T must be a Token class", std::is_convertible<T*, const Token*>)>
TokenRange<T> makeTokenRange(T* start, T* last)
{
    return TokenRange<T>{start, last};
}
```

shaneasd replied:
standard library is usually up to but not including the end right? i.e. the range is [first, last)

I did consider a factory method from which the type could be deduced but I found it didn't save that many characters over just calling the constructor. Now that there is TokenRange and ConstTokenRange to remove the template parameter there doesn't seem much point.

danmar replied:
> i.e. the range is [first, last)

That is what I understand. You can usually give the end iterator as last.

shaneasd replied:
I've decided the usefulness of until outweighs the potential ambiguity which can be mitigated with documentation so went ahead and implemented it.

pfultz2 commented on lib/tokenrange.h:
```diff
@@ -0,0 +1,65 @@
+    };
+
+    TokenIterator begin() { return TokenIterator(mFront); }
+    TokenIterator end() { return TokenIterator(mBack); }
```
These methods can be `const`.

pfultz2 commented on lib/tokenrange.h:
```diff
@@ -0,0 +1,65 @@
+        TokenIterator& operator++() { mt = mt->next(); return *this; }
+        bool operator==(const TokenIterator& b) const { return mt == b.mt; }
+        bool operator!=(const TokenIterator& b) const { return mt != b.mt; }
+        T* operator*() { return mt; }
```
This should be `const` as well.

danmar commented on test/testtokenrange.cpp:
```diff
@@ -0,0 +1,108 @@
+        const Token* expected = var.tokens()->tokAt(4);
+        const Token* end = var.tokens()->tokAt(10);
+        for (auto t : ConstTokenRange{ expected, end }) {
+            ASSERT_EQUALS(expected, t);
```
I am skeptic about comparing pointers with ASSERT_EQUALS. Imagine that the assertion fails. Then I fear that the output will look like this:

   expected: 0x12345678 actual: 0x12345670

.. no way to see which token the expected and actual are .. just that they are different. The ASSERT might be more proper.

also I believe it is unfortunate to put an assertion in a for loop body. if it fails we will see what line it fails on but not how far the testing went before there was an assertion failure. it will not be informative.

what do you think about something like:

    const Token *start = var.tokens()->tokAt(4);
    const Token *end = var.tokens()->tokAt(10);
    int count = 0;
    for (const Token *tok: ConstTokenRange{start,end})
    {
        if (tok != start->tokAt(count))
            break;
        count ++;
    }
    ASSERT_EQUALS(6, count);

shaneasd replied:
I don't think it's any worse. I'm usually not fussed about assertion messages because I figure when a test fails you're going ot debug it anyway. I do feel strongly that we should be comparing the pointers (rather than say strings) but your solution does do that.

danmar replied:
there is some truth to that. you probably want to debug it also. but well the assertion output can be pretty useful. please change it..

shaneasd replied:
I had a play with this. I'm not super happy about having to hard code the number of tokens because it's less flexible. We're trying to test the Token class not the tokenizer and there are some situations where I think the tokenizer gets it wrong which I'd like to be able to fix without updating unrelated tests like this. What are your thoughts on this alternative which I think  achieves the best of both worlds?

```
    void enumerationToEnd() const {
        givenACodeSampleToTokenize var("void a(){} void main(){ if(true){a();} }");
        const Token* expected = var.tokens();
        int correctTokens = 0;
        for (auto t : ConstTokenRange{var.tokens(), nullptr }) {
            std::ostringstream message;
            message << "Token mismatch on token " << correctTokens;
            ASSERT_EQUALS_MSG(expected, t, message.str());
            expected = t->next();
        }
        const Token* const terminator = nullptr;
        ASSERT_EQUALS(terminator, expected);
    }
```

danmar replied:
It seems we should avoid `Tokenizer` in this test. Could you try out `TokenList` instead? Something like:

        std::istringstream istr(code);
        TokenList tokenList(&settings);
        tokenList.createTokens(istr, "test.cpp");

        const Token *startToken = tokenList.front()->tokAt(123);
        ....

danmar replied:
I guess we want to refactor the test also.. create some kind of test function. For instance it could return some error message. Maybe `testTokenRange(code,start,end)` is a good signature. And then we can write a test more or less like:

    ASSERT_EQUALS("", testTokenRange("your code", 1, 10))

danmar replied:
breaking it out is maybe not required.. I do not know if there are a lot of variations to test. maybe 1-3 tests are enough.

shaneasd replied:
I agree it makes sense to use TokenList rather than Tokenizer to generate the tokens because that minimizes the dependencies of the test. I still wouldn't want to manually count the tokens though.

I think it's probably worth  refactoring out the check as you suggest.

shaneasd replied:
actually I have to use the tokenizer for the test involving scopes because I need the symbol database to generate the scopes

shaneasd replied:
take a look at the latest incarnation. The first three tests are modified in the way you suggest (more or less). The last two are sort of different so the suggestions don't apply as far as I can see.

pfultz2 commented on lib/tokenrange.h:
```diff
@@ -0,0 +1,65 @@
+        using iterator_category = std::forward_iterator_tag;
+        using value_type = T*;
+        using difference_type = std::ptrdiff_t;
+        using pointer = void;
```
It is common for things like output iterators to set the typedef to `void` when its not applicable. You can see that for [`std::back_insert_iterator`](https://en.cppreference.com/w/cpp/iterator/back_insert_iterator) as an example. Since the token iterator doesn't define `operator->`, then it seems to make sense to set this to `void` as well.

shaneasd commented:
It seems that my assertEquals overload is picking up cases where we're asserting that an integer is either true or false. This seems like a poor choice of comparison to me but maybe there's a good reason it's done that way. I think I don't actually need the overload anymore so I can delete it if necessary.

danmar commented:
> This seems like a poor choice of comparison to me but maybe there's a good reason it's done that way.

I spontaneously think so too. Not sure why we would do that. If we want that somewhere it would be better to have an explicit conversion of nonzero integer to true in the test.

danmar commented on test/testsuite.h:
```diff
@@ -29,6 +29,8 @@
 
 class options;
 
+std::string writestr(const std::string& str, bool gccStyle = false);
```
I'd prefer that `writestr` is not a global scope function. If you have to make it non-local then I think you should be able to convert it into a private method.

danmar reviewed (approved):
I feel happy about this.. if CI is happy..

shaneasd commented:
CI is not happy because of the bugs that the assertequals change has revealed. I'll work on sorting them out (probably in a new PR since it's tangential to the key change in this one).

shaneasd commented:
I've pulled in main to resolve some warnings and made a change in e2df250 to correct the use of importing constructors which is a feature not supported in c++11.

The outstanding CI failure ```scriptcheck / build (3.8) (pull_request)``` I don't really understand. Can you provide any guidance?

danmar commented:
> The outstanding CI failure scriptcheck / build (3.8) (pull_request) I don't really understand. Can you provide any guidance?

That was likely something flaky problem in the environment. Doesn't say much.

I have seen that builds and uploads can be flaky.

shaneasd commented:
well that's the only CI problem I have so I'll leave it to you to either merge or let me know what you'd like me to change.

danmar commented:
> well that's the only CI problem I have so I'll leave it to you to either merge or let me know what you'd like me to change.

I want to make it green first. But I do think this is something flaky.

danmar commented:
Could you test to run the dmake tool and see if that helps.. probably the Makefile will be updated. Possibly that is why the scriptcheck is failing.

shaneasd commented:
Hi @danmar . I'm not sure what you're asking me to do. Is the tool you mentioned distributed with cppcheck or is it a separate tool? I this it https://www.openoffice.org/tools/dmake/ ? Do you know if it's distributed in binary anywhere?

danmar commented:
it's a cppcheck tool. you can build and run it like this:

    cd cppcheck
    make dmake
    ./dmake

If you can't run it maybe we can help out with that.

shaneasd commented:
I don't make make in my path. I tried with gnumake32 and cmake and neither worked. I also noticed there's already a visual studio solution but that fails to compile because it can't find a definition of caseInsensitiveStringCompare.

So yes some assistance would be helpful thanks :)

danmar commented:
Can you fix the conflict to start with?

danmar commented:
Thanks! I have merged your changes with a5664c3e4900f46df5ad774bdc2181ebdb4dc56c

shaneasd commented:
fair enough :)
